src/michael_scott_queue.ml

```diff
@@ -40,13 +42,21 @@ let pop q =
     in
     match nhead with
     | Nil -> None
-    | Next (v, _) when Atomic.compare_and_set q.head s nhead -> Some v
+    | Next (v, x) when Atomic.compare_and_set q.head s (Next (Obj.magic (), x))
```

Comment:
Here we allocate a new node for the head.  This way we don't keep a reference to the value `v`.  The same needs to be done in `clean_until`.

An alternative would be to make nodes hold the value in a mutable field and assign `Obj.magic ()` to the field after a successful CAS.  Unfortunately that would break the snapshot functionality, because then, after calling snapshot, pop operations on the queue would destroy the snapshot.